File: scripts/diff-schemas.py
#!/usr/bin/env python3
# DEPRECATED: use crates/cfa-codegen instead. Will be removed in v0.3.
# Equivalent: `cargo run -p cfa-codegen -- diff-schemas <a> <b>`
"""Diff two `schemas.json` snapshots and classify changes per the cfa-core
schema versioning policy (plugins/cfa-core/docs/SCHEMA-VERSIONING.md).

Usage:
    python3 scripts/diff-schemas.py <ref-a> <ref-b>

Where each ref is either:
    - a git ref (e.g. `origin/main`, `HEAD`, `v1.0.0`) — the script reads
      schemas.json from that ref via `git show`.
    - a path to a schemas.json file on disk.

Output:
    Categorised diff with severity (MAJOR / MINOR / PATCH) so a CI check or a
    code reviewer can decide whether the PR's `schema_version` bump matches
    the actual change.
"""
from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent
SCHEMAS_PATH = "plugins/cfa-core/mcp/src/schemas.json"


def load(ref: str) -> dict:
    """Load schemas.json from a git ref or filesystem path."""
    p = Path(ref)
    if p.exists():
        return json.loads(p.read_text())
    try:
        out = subprocess.run(
            ["git", "show", f"{ref}:{SCHEMAS_PATH}"],
            cwd=REPO,
            capture_output=True,
            text=True,
            check=True,
        )
        return json.loads(out.stdout)
    except subprocess.CalledProcessError as e:
        print(f"error: could not read {SCHEMAS_PATH} from {ref}: {e.stderr}", file=sys.stderr)
        sys.exit(2)


def index_tools(snapshot: dict) -> dict[str, dict]:
    """Return { tool_name: schema_dict } indexed for diffing.

    The current schemas.json doesn't include the per-tool field structure
    (it lists missing entries only). For full diff we'd need to extend
    gen-zod-schemas.py to emit field-level info into the manifest. For now
    we treat each tool as a presence/absence signal plus the missing list.
    """
    tools = {}
    # Anything in the manifest's "missing" list is a passthrough — present
    # but with no typed schema.
    for entry in snapshot.get("missing", []):
        tools[entry["tool"]] = {"typed": False, "struct": entry["struct"]}
    # The manifest doesn't currently list the typed tools by name. Until
    # gen-zod-schemas.py is extended, fall back to inspecting schemas.ts.
    return tools


def classify(a: dict, b: dict) -> dict[str, list[str]]:
    """Return { severity: [reasons] }."""
    out: dict[str, list[str]] = {"MAJOR": [], "MINOR": [], "PATCH": []}

    a_tools = set(a.get("tools", []))
    b_tools = set(b.get("tools", []))

    added = sorted(b_tools - a_tools)
    removed = sorted(a_tools - b_tools)
    if added:
        sample = ", ".join(added[:5]) + (" ..." if len(added) > 5 else "")
        out["MINOR"].append(f"+{len(added)} tools (e.g. {sample})")
    if removed:
        sample = ", ".join(removed[:5]) + (" ..." if len(removed) > 5 else "")
        out["MAJOR"].append(f"−{len(removed)} tools removed (e.g. {sample})")

    a_missing = {entry["tool"] for entry in a.get("missing", [])}
    b_missing = {entry["tool"] for entry in b.get("missing", [])}
    newly_passthrough = sorted(b_missing - a_missing)
    newly_typed = sorted(a_missing - b_missing)
    if newly_passthrough:
        out["MAJOR"].append(
            f"became passthrough: {', '.join(newly_passthrough)} — "
            "regression: a previously-typed schema lost its types"
        )
    if newly_typed:
        out["MINOR"].append(f"became typed: {', '.join(newly_typed)}")

    a_ver = a.get("schema_version", "?")
    b_ver = b.get("schema_version", "?")
    if a_ver != b_ver:
        out["PATCH"].append(f"schema_version field: {a_ver} → {b_ver}")

    return out


def main() -> int:
    if len(sys.argv) != 3:
        print(__doc__, file=sys.stderr)
        return 2

    a = load(sys.argv[1])
    b = load(sys.argv[2])

    print(f"Comparing {sys.argv[1]} → {sys.argv[2]}")
    print()

    classified = classify(a, b)

    if not any(classified.values()):
        print("No structural changes detected.")
        return 0

    for severity in ("MAJOR", "MINOR", "PATCH"):
        reasons = classified[severity]
        if not reasons:
            continue
        print(f"## {severity}")
        for r in reasons:
            print(f"  - {r}")
        print()

    if classified["MAJOR"]:
        print("Recommended schema_version bump: MAJOR")
        return 0
    if classified["MINOR"]:
        print("Recommended schema_version bump: MINOR")
        return 0
    print("Recommended schema_version bump: PATCH")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/cfa-codegen/src/commands/diff_schemas.rs
//! `cfa-codegen diff-schemas <a> <b>` — port of `scripts/diff-schemas.py`.
//!
//! Compares two `schemas.json` snapshots (loaded from a filesystem path or
//! a git ref via `git show`) and classifies the differences by semver
//! severity per the cfa-core schema versioning policy.

use std::path::{Path, PathBuf};
use std::process::Command;

use anyhow::{anyhow, Context, Result};
use serde_json::Value;

const SCHEMAS_PATH: &str = "plugins/cfa-core/mcp/src/schemas.json";

/// Load a schemas.json snapshot from a filesystem path or git ref.
pub fn load(repo_root: &Path, reference: &str) -> Result<Value> {
    let p = PathBuf::from(reference);
    if p.exists() {
        let text = std::fs::read_to_string(&p)
            .with_context(|| format!("read {}", p.display()))?;
        return serde_json::from_str(&text)
            .with_context(|| format!("parse {}", p.display()));
    }
    // Treat as git ref.
    let output = Command::new("git")
        .arg("show")
        .arg(format!("{}:{}", reference, SCHEMAS_PATH))
        .current_dir(repo_root)
        .output()
        .with_context(|| format!("run git show {}:{}", reference, SCHEMAS_PATH))?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(anyhow!(
            "could not read {} from {}: {}",
            SCHEMAS_PATH,
            reference,
            stderr
        ));
    }
    let text = String::from_utf8(output.stdout)?;
    serde_json::from_str(&text).with_context(|| format!("parse {} from {}", SCHEMAS_PATH, reference))
}

#[derive(Default)]
pub struct Classified {
    pub major: Vec<String>,
    pub minor: Vec<String>,
    pub patch: Vec<String>,
}

impl Classified {
    pub fn is_empty(&self) -> bool {
        self.major.is_empty() && self.minor.is_empty() && self.patch.is_empty()
    }
}

pub fn classify(a: &Value, b: &Value) -> Classified {
    let mut out = Classified::default();

    let a_tools: std::collections::BTreeSet<String> = a
        .get("tools")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();
    let b_tools: std::collections::BTreeSet<String> = b
        .get("tools")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();

    let added: Vec<&String> = b_tools.difference(&a_tools).collect();
    let removed: Vec<&String> = a_tools.difference(&b_tools).collect();
    if !added.is_empty() {
        let sample = sample(&added);
        out.minor
            .push(format!("+{} tools (e.g. {})", added.len(), sample));
    }
    if !removed.is_empty() {
        let sample = sample(&removed);
        out.major.push(format!(
            "−{} tools removed (e.g. {})",
            removed.len(),
            sample
        ));
    }

    let a_missing = missing_tool_set(a);
    let b_missing = missing_tool_set(b);
    let newly_passthrough: Vec<&String> = b_missing.difference(&a_missing).collect();
    let newly_typed: Vec<&String> = a_missing.difference(&b_missing).collect();
    if !newly_passthrough.is_empty() {
        let names: Vec<String> = newly_passthrough.iter().map(|s| (*s).clone()).collect();
        out.major.push(format!(
            "became passthrough: {} — regression: a previously-typed schema lost its types",
            names.join(", ")
        ));
    }
    if !newly_typed.is_empty() {
        let names: Vec<String> = newly_typed.iter().map(|s| (*s).clone()).collect();
        out.minor.push(format!("became typed: {}", names.join(", ")));
    }

    let a_ver = a.get("schema_version").and_then(|v| v.as_str()).unwrap_or("?");
    let b_ver = b.get("schema_version").and_then(|v| v.as_str()).unwrap_or("?");
    if a_ver != b_ver {
        out.patch
            .push(format!("schema_version field: {} → {}", a_ver, b_ver));
    }

    out
}

fn missing_tool_set(snapshot: &Value) -> std::collections::BTreeSet<String> {
    snapshot
        .get("missing")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|entry| entry.get("tool").and_then(|t| t.as_str()).map(String::from))
                .collect()
        })
        .unwrap_or_default()
}

fn sample(items: &[&String]) -> String {
    let head: Vec<&str> = items.iter().take(5).map(|s| s.as_str()).collect();
    let mut s = head.join(", ");
    if items.len() > 5 {
        s.push_str(" ...");
    }
    s
}

pub fn run(repo_root: &Path, ref_a: &str, ref_b: &str) -> Result<i32> {
    let a = load(repo_root, ref_a)?;
    let b = load(repo_root, ref_b)?;

    println!("Comparing {} → {}", ref_a, ref_b);
    println!();

    let classified = classify(&a, &b);
    if classified.is_empty() {
        println!("No structural changes detected.");
        return Ok(0);
    }

    for (label, reasons) in [
        ("MAJOR", &classified.major),
        ("MINOR", &classified.minor),
        ("PATCH", &classified.patch),
    ] {
        if reasons.is_empty() {
            continue;
        }
        println!("## {}", label);
        for r in reasons {
            println!("  - {}", r);
        }
        println!();
    }

    if !classified.major.is_empty() {
        println!("Recommended schema_version bump: MAJOR");
    } else if !classified.minor.is_empty() {
        println!("Recommended schema_version bump: MINOR");
    } else {
        println!("Recommended schema_version bump: PATCH");
    }
    Ok(0)
}
